### PR TITLE
Add support for storing AoC related data in site

### PR DESCRIPTION
**Pull request by D0rs4n**
## Summary

I added support for storing AoC related data in site

## Description of changes

Added two new endpoints, with relating ViewSets and models

### I confirm I have:

- [x] Joined the [Python Discord community](discord.gg/python)
- [x] Read the [Code of Conduct](https://www.pydis.com/pages/code-of-conduct) and agree to it
- [x] I have discussed implementing this feature on the relevant service (Discord, GitHub, etc.)

### CODE REVIEW

File: pydis_site/apps/api/models/bot/aoc_completionist_block.py
Comment (jchristgit) [inline pydis_site/apps/api/models/bot/aoc_completionist_block.py:22]: Wouldn't every user in this table be blocked? Should we add / use a reason field instead?

File: pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py
Comment (jchristgit) [inline pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py]: nit:
```suggestion
    >>> {
    ...     'user': int,
    ...     'is_blocked': bool
    ... }
```

File: pydis_site/apps/api/viewsets/bot/aoc_link.py
Comment (jchristgit) [inline pydis_site/apps/api/viewsets/bot/aoc_link.py]: Nit:
```suggestion
    ...     'user': int,
    ...     'aoc_username': str
```

**Review by jchristgit — APPROVED** [review]
One minor question and a few nitpicks. The question is probably not as useful. Thanks!

File: pydis_site/apps/api/models/bot/aoc_completionist_block.py
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/aoc_completionist_block.py]: This should probably be a `OneToOneField` instead. A many `AocCompletionistBlock`-to-one `User` relationship does not make sense since each user can only be blocked once.

File: pydis_site/apps/api/models/bot/aoc_completionist_block.py
Comment (MarkKoz) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py:22]: Yeah, I think it can be assumed the user is blocked by virtue of just existing in this table. Admittedly, it is kind of strange to have a table with just one field but on the other hand `is_blocked` seems redundant.

File: pydis_site/apps/api/models/bot/aoc_link.py
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/aoc_link.py]: This should be a `OneToOneField`. I realise this is using the Discord User model rather than the Django user model, but I believe this is still applicable https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#extending-the-existing-user-model

File: pydis_site/apps/api/models/bot/aoc_link.py
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/aoc_link.py:8]: Can only linked users be blocked from being a completionist? Is that separate from this table because we need to be able to keep a user blocked if they unlink their account?

If that is the case, it may be better to merge the tables, and instead of deleting the record when the user unlinks, just set `aoc_username` to null/empty. This way, the block will remain for unlinked users. What are everyone's thoughts on this alternative?

**Review by MarkKoz — CHANGES_REQUESTED** [review]

File: pydis_site/apps/api/models/bot/aoc_completionist_block.py
Comment (D0rs4n) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py:22]: I see the concern, however there is no reason provided on the bot side, as a golden mean I could implement a deactivate block command which sets the `is_blocked` field to False whilst keeping a record of users who have been blocked previously.

File: pydis_site/apps/api/models/bot/aoc_completionist_block.py
Comment (D0rs4n) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py]: Oh, yes. I didn't catch it. Thanks!

File: pydis_site/apps/api/models/bot/aoc_link.py
Comment (D0rs4n) [reply pydis_site/apps/api/models/bot/aoc_link.py:8]: No, I don't believe so, this way Admins can block users even if they haven't linked their AoC account.
Otherwise, they would be limited to wait until they link their account, and then block them.
Update: Sorry, I see your alternative now. I was trying to keep the tables as close to the code as possible, as they are two unlinked commands

File: pydis_site/apps/api/models/bot/aoc_completionist_block.py
Comment (janine9vn) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py:22]: It might make sense to keep the column in case we need to unblock someone. Altering a value in a row is ever so slightly easier than deleting a row. I also wouldn't be opposed to a reason field, so there's some context around it.

Comment (D0rs4n) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py:22]: Alright, that might need some modification on the bot site, although.
But, minor ones, only.

Comment (D0rs4n) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py:22]: Added in b93dce5

**Review by jchristgit — APPROVED** [review]
⭐⭐⭐⭐⭐⭐

File: pydis_site/apps/api/models/bot/aoc_completionist_block.py
Comment (MarkKoz) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py:22]: If we want to keep a reason around, then the `is_blocked` field makes more sense. However, this is starting to look more like an infraction at this point.

Comment (D0rs4n) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py:22]: It might look like one, but admins might not want to block someone and give them an infraction the same time.
Especially, because that's on Events Bot, (used to be on Sir-Lanebot)

File: pydis_site/apps/api/models/bot/aoc_link.py
Comment (MarkKoz) [reply pydis_site/apps/api/models/bot/aoc_link.py:8]: > I was trying to keep the tables as close to the code as possible

I'm not sure if I agree with this approach going forward with other migrations from redis to the API. The design that makes sense for redis doesn't necessarily make sense for the API. I'd rather not have the API be beholden to the choices that were made for redis. This is more of a general comment; don't take it to mean that my alternative suggestion in this case was definitely the better choice.

If we want to make these migrations quick, then that's okay. But we should revisit it eventually then.

Comment (D0rs4n) [reply pydis_site/apps/api/models/bot/aoc_link.py:8]: Regardless of the general approach proposed, I'd still part with using two tables, to keep things consistent, and easier to handle on the bot side, as they're two distinct commands.

File: pydis_site/apps/api/models/bot/aoc_completionist_block.py
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/aoc_completionist_block.py:24]: Better to use `blank` instead of `null`. But maybe we should leave it `null` since that seems to be what other endpoints use for similar fields (and change them all later in a separate PR if we want)?

https://docs.djangoproject.com/en/3.2/ref/models/fields/#null

Comment (D0rs4n) [reply pydis_site/apps/api/models/bot/aoc_completionist_block.py:24]: Yes, I used Infractions (the model) as a reference, to keep things consistent

File: pydis_site/apps/api/serializers.py
Comment (MarkKoz) [inline pydis_site/apps/api/serializers.py:262]: I'm not sure, but I think this is redundant since it's listing all fields. I believe leaving the fields unspecified would be equivalent.

File: pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py]:
```suggestion
    ...         "is_blocked": False,
```

File: pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py]:
```suggestion
    ...         "is_blocked": False,
```

File: pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py]:
```suggestion
    ...     "is_blocked": bool,
```

File: pydis_site/apps/api/serializers.py
Comment (D0rs4n) [reply pydis_site/apps/api/serializers.py:262]: Doesn't this makes it a bit easier to understand?
In my personal opinion it makes it easier to understand what those serializers are for, and that's our aim I believe.

File: pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py]: Probably want to be able to filter by `is_blocked` too.

File: pydis_site/apps/api/serializers.py
Comment (MarkKoz) [reply pydis_site/apps/api/serializers.py:262]: I don't know. Maybe. But it's also yet another thing to maintain when a field gets changed. On the other hand, maybe we don't want new fields to implicitly be included in the serializer. Just a comment so I don't feel strongly either way.

File: pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_completionist_block.py]:
```suggestion
    Deletes the AoC Completionist block item with the given `user__id`.

    #### Status codes
```

File: pydis_site/apps/api/viewsets/bot/aoc_link.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_link.py]:
```suggestion
    Deletes the AoC account link item with the given `user__id`.

    #### Status codes
```

File: pydis_site/apps/api/viewsets/bot/aoc_link.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_link.py]: Does this really still return a list for just 1 user? I am pretty sure the default DRF behaviour is to return a single object for /endpoint/id requests.

File: pydis_site/apps/api/viewsets/bot/aoc_link.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_link.py]: Surely it's meant to be this?
```suggestion
    ### GET /bot/aoc-account-links/<user__id:int>
```

File: pydis_site/apps/api/viewsets/bot/aoc_link.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_link.py]: Some minor things in the docs like `user__id` not always being surrounded by backticks or some sentences not having full stops, but I will leave it up to you if you want to be bothered with fixing that.

Comment (MarkKoz) [reply pydis_site/apps/api/viewsets/bot/aoc_link.py]: Another thing is inconsistent use of tense/mood.

File: pydis_site/apps/api/viewsets/bot/aoc_link.py
Comment (D0rs4n) [reply pydis_site/apps/api/viewsets/bot/aoc_link.py]: Yes, I was under the impression that it does, but after testing it out yesterday, it seemed to return a list still.

Comment (D0rs4n) [reply pydis_site/apps/api/viewsets/bot/aoc_link.py]: Don't mind me, you're right! :D 
Not sure why I updated in the end.

File: pydis_site/apps/api/viewsets/bot/aoc_link.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_link.py]: This was correct originally.
```suggestion
    - 204: returned on success
```

**Review by MarkKoz — CHANGES_REQUESTED** [review]

File: pydis_site/apps/api/viewsets/bot/aoc_link.py
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/aoc_link.py]: Shouldn't be indented as far.

**Review by MarkKoz — CHANGES_REQUESTED** [review]

**Comment by MarkKoz** [comment]
Are you not planning on unit tests by the way? Not sure if there is much to gain anyway since this is mostly relying on DRF.

**Comment by D0rs4n** [comment]
Initially I was not, I checked out a similar migration PR (namely the PR made by @ChrisLovering ) to see if there were tests.
But yes, these things are strongly reliant on DRF, and I did not introduce things such as complicated constraints, or custom actions.

**Review by MarkKoz — APPROVED** [review]

**Pull request merged**
